Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
+
+    case google_firestore_v1_StructuredQuery_FieldFilter_Operator_IN:
+      return Filter::Operator::In;
+
+    case google_firestore_v1_StructuredQuery_FieldFilter_Operator_ARRAY_CONTAINS_ANY:  // NOLINT
+      return Filter::Operator::ArrayContainsAny;
+
+    default:
+      reader->Fail(StringFormat("Unhandled FieldFilter.op: %s", op));
+      return Filter::Operator{};
+  }
+}
+
+google_firestore_v1_StructuredQuery_Order* Serializer::EncodeOrderBys(
+    const OrderByList& orders) const {
+  auto* result = MakeArray<google_firestore_v1_StructuredQuery_Order>(
+      CheckedSize(orders.size()));
+
+  int i = 0;
+  for (const OrderBy& order : orders) {
+    auto& encoded_order = result[i];
+
+    encoded_order.field.field_path = EncodeFieldPath(order.field());
+    auto dir = order.ascending()
+                   ? google_firestore_v1_StructuredQuery_Direction_ASCENDING
+                   : google_firestore_v1_StructuredQuery_Direction_DESCENDING;
+    encoded_order.direction = dir;
+
+    ++i;
+  }
+
+  return result;
+}
+
+OrderByList Serializer::DecodeOrderBys(
+    nanopb::Reader* reader,
+    google_firestore_v1_StructuredQuery_Order* order_bys,
+    pb_size_t size) const {
+  OrderByList result;
+  result = result.reserve(size);
+
+  for (pb_size_t i = 0; i != size; ++i) {
+    result = result.push_back(DecodeOrderBy(reader, order_bys[i]));
+  }
+
+  return result;
+}
+
+OrderBy Serializer::DecodeOrderBy(
+    nanopb::Reader* reader,
+    const google_firestore_v1_StructuredQuery_Order& order_by) const {
+  auto field_path =
+      FieldPath::FromServerFormat(DecodeString(order_by.field.field_path));
+
+  Direction direction;
+  switch (order_by.direction) {
+    case google_firestore_v1_StructuredQuery_Direction_ASCENDING:
+      direction = Direction::Ascending;
+      break;
+
+    case google_firestore_v1_StructuredQuery_Direction_DESCENDING:
+      direction = Direction::Descending;
+      break;
+
+    default:
+      reader->Fail(StringFormat(
+          "Unrecognized google_firestore_v1_StructuredQuery_Direction %s",
+          order_by.direction));
+      return OrderBy{};
+  }
+
+  return OrderBy(std::move(field_path), direction);
+}
+
+google_firestore_v1_Cursor Serializer::EncodeBound(const Bound& bound) const {
+  google_firestore_v1_Cursor result{};
+  result.before = bound.before();
+
+  auto count = CheckedSize(bound.position().size());
+  result.values_count = count;
+  result.values = MakeArray<google_firestore_v1_Value>(count);
+
+  int i = 0;
+  for (const FieldValue& field_value : bound.position()) {
+    result.values[i] = EncodeFieldValue(field_value);
+    ++i;
+  }
+
+  return result;
+}
+
+std::shared_ptr<Bound> Serializer::DecodeBound(
+    nanopb::Reader* reader, const google_firestore_v1_Cursor& cursor) const {
+  std::vector<FieldValue> index_components;
+  index_components.reserve(cursor.values_count);
+
+  for (pb_size_t i = 0; i != cursor.values_count; ++i) {
```

Comment:
More conventionally this is written `i < cursor.values_count`.

Reply:
Interesting perspective.

From my point of view less-than is better because it decouples loop termination from the step. It essentially declares for which values of the loop index the loop is valid without regard for how exactly we're stepping. In this case we're expressing that we should not go past the end of the array.

This has the effect of making the loop potentially more resilient to change. For example, should someone come along later and add a special case where the loop index is bumped an extra step to exclude the next value, we wouldn't need to specially handle the boundary condition where the last entry prompted that. Depending on the use case that may or may not be valid, but at least you're not going to find out about it with an infinite loop (or in this case a crash after reading past the end of mapped memory).

In short, I think it's reasonable to write these loops as if the loop variable may be increased by more than one precisely because someone may come along and make a change to do just that.

Note: I'm mostly sharing perspective here rather than advocating that you necessarily make the change. If you agree, great. If not, feel free to ignore.